near-sdk/src/types/gas.rs

```diff
@@ -0,0 +1,115 @@
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+/// Represents the amount of NEAR tokens in "gas units" which are used to fund transactions.
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    BorshSerialize,
+    BorshDeserialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct Gas(u64);
+
+impl Gas {
+    pub const fn new(amount: u64) -> Self {
+        Self(amount)
+    }
+}
+
+impl Serialize for Gas {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.0.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for Gas {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        s.parse::<u64>().map(Self).map_err(|err| de::Error::custom(err.to_string()))
+    }
+}
+
+impl From<u64> for Gas {
+    fn from(amount: u64) -> Self {
+        Self(amount)
+    }
+}
+
+impl From<Gas> for u64 {
+    fn from(gas: Gas) -> Self {
+        gas.0
+    }
+}
+
+impl core::ops::Add for Gas {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        Self(self.0 + other.0)
+    }
+}
+
+impl core::ops::AddAssign for Gas {
+    fn add_assign(&mut self, other: Self) {
+        self.0 += other.0;
+    }
+}
+
+impl core::ops::SubAssign for Gas {
+    fn sub_assign(&mut self, other: Self) {
+        self.0 -= other.0;
+    }
+}
+
+impl core::ops::Sub for Gas {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        Self(self.0 - other.0)
+    }
+}
+
+impl core::ops::Add<u64> for Gas {
+    type Output = Self;
+
+    fn add(self, other: u64) -> Self {
+        Self(self.0 + other)
+    }
+}
+
+impl core::ops::Sub<u64> for Gas {
+    type Output = Self;
+
+    fn sub(self, other: u64) -> Self {
+        Self(self.0 - other)
+    }
+}
+
+impl core::ops::Mul<u64> for Gas {
+    type Output = Self;
+
+    fn mul(self, other: u64) -> Self {
+        Self(self.0 * other)
+    }
+}
+
+impl core::ops::Div<u64> for Gas {
+    type Output = Self;
+
+    fn div(self, other: u64) -> Self {
+        Self(self.0 / other)
+    }
```

Comment:
Feels a bit strange to provide `div` without also providing `div_rem` -- division is a lossy operation (it forgets remined), for orthogonal API there shoud be a way to recover the rem.

Reply:
Do you mean just implementing `Rem` or implementing some trait from `num_integer` or something similar. I'm not sure what `div_rem` you are referring to, unless you mean as a core impl

Reply:
Hm, I was sure std had `div_rem` function, but that's not the case actually. Yeah, in this case just `Rem` would do the trick.

Reply:
Added, just couldn't think of a reason that `Rem` would be needed for usage with Gas.

Also, just to confirm, you also think it makes the most sense to have the `Mul`, `Div`, and `Rem` operators be on `u64` rather than gas? The asymmetry is a bit weird to me, but in practice it seems to make more sense

Reply:
There's no assymetry here -- for mul, div and rem, the multiplier is [dimensionless](https://en.wikipedia.org/wiki/Dimensionless_quantity). Had we used `Gas` there, we'd have to return `Gas²`.

>Added, just couldn't think of a reason that Rem would be needed for usage with Gas.

What's the use-case for Div? Div is somewhat rarely useful without Rem -- if you want to devide something in parts, you have to do something about the remainder.